external/kafka-0-10-token-provider/src/main/scala/org/apache/spark/kafka010/KafkaDelegationTokenProvider.scala

```diff
@@ -38,24 +37,49 @@ private[spark] class KafkaDelegationTokenProvider
       sparkConf: SparkConf,
       creds: Credentials): Option[Long] = {
     try {
-      logDebug("Attempting to fetch Kafka security token.")
-      val (token, nextRenewalDate) = KafkaTokenUtil.obtainToken(sparkConf)
-      creds.addToken(token.getService, token)
-      return Some(nextRenewalDate)
+      var lowestNextRenewalDate: Option[Long] = None
+      KafkaTokenSparkConf.getAllClusterConfigs(sparkConf).foreach { clusterConf =>
+        if (delegationTokensRequired(clusterConf)) {
+          try {
+            logDebug(
+              s"Attempting to fetch Kafka security token for cluster ${clusterConf.identifier}.")
+            val (token, nextRenewalDate) = KafkaTokenUtil.obtainToken(sparkConf, clusterConf)
+            creds.addToken(token.getService, token)
+            if (lowestNextRenewalDate.isEmpty || nextRenewalDate < lowestNextRenewalDate.get) {
+              lowestNextRenewalDate = Some(nextRenewalDate)
+            }
+          } catch {
+            case NonFatal(e) =>
+              logWarning(s"Failed to get token from service: $serviceName " +
+                s"cluster: ${clusterConf.identifier}", e)
+          }
+        } else {
+          logDebug(
+            s"Cluster ${clusterConf.identifier} does not require delegation token, skipping.")
+        }
+      }
+      lowestNextRenewalDate
     } catch {
       case NonFatal(e) =>
-        logWarning(s"Failed to get token from service $serviceName", e)
+        logWarning(s"Failed to get token cluster configuration", e)
+        None
```

Comment:
This isn't exactly correct, since you may get a few delegation tokens. I guess the question here is what should be the error behavior? Seems like the previous behavior is not just log the error and let the application fail later because of auth errors.

So the correct thing here seems to be to move the error handling above, where the code tries to get a token for a specific cluster, and keep going if that particular one fails.

That would also avoid the two levels of error catching you currently have.

Reply:
Having 2 level of error catching is added for the following reasons:
* Outer catch protects delegation token manager from exception coming from for example `getAllClusterConfigs`
* Inner catch protects the inner loop and just skips a particular cluster

What I think is nice catch that `delegationTokensRequired` is not covered by the inner block. As a result in the first token obtained properly but the second throws exception in `delegationTokensRequired` then the whole function will return `None`.

> That would also avoid the two levels of error catching you currently have.

 Don't know how this could work with one level of catch block. Such case if first fails the rest won't be processed. I've modified the code and share your thoughts.